File: python/data_structures/ordered_set.py
from bisect import bisect_left, bisect_right, insort


class OrderedSet:
    def __init__(self, values=()):
        self.a = sorted(set(values))

    def add(self, x):
        i = bisect_left(self.a, x)
        if i == len(self.a) or self.a[i] != x:
            self.a.insert(i, x)

    def discard(self, x):
        i = bisect_left(self.a, x)
        if i < len(self.a) and self.a[i] == x:
            self.a.pop(i)

    def order_of_key(self, x):
        return bisect_left(self.a, x)

    def find_by_order(self, k):
        return self.a[k]

File: cpp/data_structures/ordered_set.cpp
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace std;
using namespace __gnu_pbds;
template <class T>
using ordered_set = tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
template <class T> struct ordered_multiset {
  using P = pair<T, int>;
  tree<P, null_type, less<P>, rb_tree_tag, tree_order_statistics_node_update> tr;
  int timer = 0;
  void insert(T x) {
    tr.insert({x, timer++});
  }
  bool erase_one(T x) {
    auto it = tr.lower_bound({x, -1});
    if (it == tr.end() || it->first != x)
      return 0;
    tr.erase(it);
    return 1;
  }
  int order_of_key(T x) {
    return tr.order_of_key({x, -1});
  }
  T kth(int k) {
    return tr.find_by_order(k)->first;
  }
};
